mypy_django_plugin/transformers/managers.py

```diff
@@ -183,112 +247,198 @@ def create_new_manager_class_from_from_queryset_method(ctx: DynamicClassDefConte
+
+    assert isinstance(queryset_info, TypeInfo)
+
+    manager_class_name = manager_base.name + "From" + queryset_info.name
+    current_module = semanal_api.modules[semanal_api.cur_mod_id]
+    existing_sym = current_module.names.get(manager_class_name)
+    if (
+        existing_sym is not None
+        and isinstance(existing_sym.node, TypeInfo)
+        and existing_sym.node.has_base(fullnames.MANAGER_CLASS_FULLNAME)
+        and existing_sym.node.metadata.get("django", {}).get("from_queryset_manager") == queryset_info.fullname
+    ):
+        # Reuse an identical, already generated, manager
+        new_manager_info = existing_sym.node
+    else:
+        manager_base_instance = fill_typevars(manager_base)
+        assert isinstance(manager_base_instance, Instance)
+        new_manager_info = helpers.add_new_class_for_module(
+            module=current_module,
+            name=manager_class_name,
+            bases=[manager_base_instance],
+        )
+        new_manager_info.type_vars = manager_base.type_vars
+        new_manager_info.line = ctx.call.line
+        new_manager_info.defn.type_vars = manager_base.defn.type_vars
+        new_manager_info.defn.line = ctx.call.line
+
+        merge_queryset_into_manager(
+            api=semanal_api,
+            manager_info=new_manager_info,
+            queryset_info=queryset_info,
+            manager_base=manager_base,
+            class_name=manager_class_name,
         )
 
-    # Insert the new manager (dynamic) class
-    assert semanal_api.add_symbol_table_node(ctx.name, SymbolTableNode(GDEF, new_manager_info, plugin_generated=True))
+    # Whenever `<QuerySet>.as_manager()` isn't called at class level, we want to ensure
+    # that the variable is an instance of our generated manager. Instead of the return
+    # value of `.as_manager()`. Though model argument is populated as `Any`.
+    # `transformers.models.AddManagers` will populate a model's manager(s), when it
+    # finds it on class level.
+    var = Var(name=ctx.name, type=Instance(new_manager_info, [AnyType(TypeOfAny.from_omitted_generics)]))
+    var.info = new_manager_info
+    var._fullname = f"{current_module.fullname}.{ctx.name}"
+    var.is_inferred = True
+    # Note: Order of `add_symbol_table_node` calls matters. Depending on what level
+    # we've found the `.as_manager()` call. Point here being that we want to replace the
+    # `.as_manager` return value with our newly created manager.
+    assert semanal_api.add_symbol_table_node(
+        ctx.name, SymbolTableNode(semanal_api.current_symbol_kind(), var, plugin_generated=True)
+    )
+    assert semanal_api.add_symbol_table_node(
+        # We'll use `new_manager_info.name` instead of `manager_class_name` here
+        # to handle possible name collisions, as it's unique.
+        new_manager_info.name,
+        # Note that the generated manager type is always inserted at module level
+        SymbolTableNode(GDEF, new_manager_info, plugin_generated=True),
```

Comment:
```suggestion
        SymbolTableNode(semanal_api.current_symbol_kind(), new_manager_info, plugin_generated=True),
```

Reply:
The reason behind `GDEF` here is to get a fullname on the type that `AddManagers` can do lookup on. And `AddManagers` expects the type to be on module level. Otherwise we'll get types looking like `myapp.models.MyModel.ManagerFromMyQuerySet` and `AddManagers` won't find them

Reply:
Does it actually work though? Cause I observed that when I did something similar in #1045 that it would fail to look up the name because it's added inside the class.

Since this isn't actually returning a type I'm wondering if it would be better to implement this support as I've done with `.from_queryset()` in #1045 (do it in the model hook, that way mypy doesn't care about the generated type) rather than through the `dynamic_class_hook`?

Reply:
It works I think? At least through the tests that I added. I also tried to break it similarly to what happened with `.from_queryset` in https://github.com/ljodal/django-stubs-regression-test, but never got it breaking.

Perhaps it might be preferable to align with #1045 but either this PR or #1045 merges first. Instead of merging them together, as that changeset feels a bit large.

Reply:
Yeah, absolutely agree, they should be merged separately. Sounds like a good plan and if it doesn't break it's probably fine :)